File: Difficulty: Medium/Connected Components in an Undirected Graph/connected-components-in-an-undirected-graph.cpp

#define adj_t vector<vector<int>>
#define visit_t vector<bool>

class Solution {
  public:
  
    void dfs(adj_t& graph, int u, vector<bool>& visited, vector<int>& res) {
        visited[u] = true;
        res.push_back(u);
        for(int i  = 0;i<graph[u].size();i++) {
            int v = graph[u][i];
            if(!visited[v]) {
                dfs(graph,v,visited,res);
            }
        }
    }
    vector<vector<int>> getComponents(int V, vector<vector<int>>& edges) {
        // code here
        int n = edges.size();
        visit_t visited(V,false);
        adj_t graph(V);
        for(int i = 0; i<n;i++) {
            graph[edges[i][0]].push_back(edges[i][1]);
            graph[edges[i][1]].push_back(edges[i][0]);
            
        }
        
        vector<vector<int> > ans;
        
        for(int i  = 0;i<V;i++) {
            if(!visited[i]) {
                vector<int> res;
                dfs(graph,i,visited,res);
                ans.push_back(res);
            }
        }
        return ans;
    }
};

File: Difficulty: Medium/Connected Components in an Undirected Graph/connected-components-in-an-undirected-graph.java
class Solution {
    
    private void bfs(ArrayList<ArrayList<Integer>> adjlist,
                        boolean[] visited,int i, 
                        ArrayList<ArrayList<Integer>> ans) {
                            
        ArrayList<Integer> ans1 = new ArrayList<>();
        
        visited[i] = true;
        
        Deque<Integer> dq = new ArrayDeque<>();
        dq.add(i);
        
        while(dq.isEmpty()==false) {
            
            int u = dq.poll();
            ans1.add(u);
            
            for(Integer v: adjlist.get(u)) {
                if(visited[v]==false) {
                    visited[v] = true;
                    dq.add(v);
                    
                }
            }
        }
        
        ans.add(ans1);
        
    }
    
    public ArrayList<ArrayList<Integer>> getComponents(int V, int[][] edges) {
        // code here
        int n = edges.length;
        ArrayList<ArrayList<Integer>> adjlist = new ArrayList<>();
        
        for(int i = 0;i<V;i++) adjlist.add(new ArrayList<Integer>());
        
        for(int i = 0;i<n;i++) {
            int u = edges[i][0];
            int v = edges[i][1];
            
            adjlist.get(u).add(v);
            adjlist.get(v).add(u);
        }
        
        ArrayList<ArrayList<Integer>> ans = new ArrayList<>();
        boolean[] visited = new boolean[V];
        
        for(int i = 0;i<V;i++) {
            if(visited[i]==false) {
                bfs(adjlist,visited,i,ans);
            }
        }
        
        return ans;
        
    }
}